src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1419,6 +1595,266 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
     setFileStats();
 }
 
+std::uint32_t
+DatabaseShardImp::getShardBoundaryIndex(std::lock_guard<std::mutex> const&)
+{
+    auto const validIndex = app_.getLedgerMaster().getValidLedgerIndex();
+
+    // Shards with an index earlier than recentShardBoundaryIndex
+    // are considered historical. The three shards at or later than
+    // this index consist of the two most recently validated shards
+    // and the shard still in the process of being built by live
+    // transactions.
+    return NodeStore::seqToShardIndex(validIndex, ledgersPerShard_) - 1;
+}
+
+std::uint32_t
+DatabaseShardImp::countHistoricalShards(std::lock_guard<std::mutex> const& lock)
+{
+    auto const recentShardBoundaryIndex = getShardBoundaryIndex(lock);
+
+    return std::count_if(
+        shards_.begin(),
+        shards_.end(),
+        [recentShardBoundaryIndex](decltype(shards_)::value_type const& entry) {
+            return entry.first < recentShardBoundaryIndex;
+        });
+}
+
+void
+DatabaseShardImp::checkRecentShards(std::lock_guard<std::mutex> const& lock)
+{
+    if (auto& cur = latestShardIndex_, &prev = secondLatestShardIndex_;
+        cur || prev)
+    {
+        auto const latestShardIndex = NodeStore::seqToShardIndex(
+            app_.getLedgerMaster().getValidLedgerIndex(), ledgersPerShard_);
+
+        auto const separateHistoricalPath = !historicalPaths_.empty();
+
+        //////////////////////////////////////////////////////////////////
+        // Lambdas used inside of this method:
+        auto const removeShard = [this](std::uint32_t const shardIndex) {
```

Comment:
Consider adding return types for all these lambdas. We often omit return types, for when there are several helper functions like this I'd prefer to add them.

Reply:
Done